Source/Softswitch/src/softswitch_common.cpp

```diff
@@ -3,383 +3,669 @@
+ * softswitch_onReceive: Executed any time a message is received.
+ *----------------------------------------------------------------------------*/
+void softswitch_onReceive(ThreadCtxt_t* ThreadContext, volatile void* recv_buf)
 {
-    uint32_t cur_device = thr_ctxt->nextOnIdle & ~P_ONIDLE_CHANGE;
-    if (cur_device >= thr_ctxt->numDevInsts) return false;
-    uint32_t last_device = cur_device+IDLE_SWEEP_CHUNK_SIZE;
-    if (last_device >= (thr_ctxt->numDevInsts)) last_device = IDLE_SWEEP_CHUNK_SIZE - thr_ctxt->numDevInsts + last_device;
-    while (cur_device != last_device)
+    devInst_t* recvDevBegin;
+    devInst_t* recvDevEnd;
+    
+    ThreadContext->rxCount++;                  // Increment received message count
+    
+    // Grab the message header 
+    volatile P_Msg_Hdr_t* recvHdr = static_cast<volatile P_Msg_Hdr_t*>(recv_buf);
+    
+    // Decode the header
+    uint32_t devAdr = (recvHdr->swAddr & P_SW_DEVICE_MASK) >> P_SW_DEVICE_SHIFT;
+    uint8_t opcode = (recvHdr->swAddr & P_SW_OPCODE_MASK) >> P_SW_OPCODE_SHIFT;
+    uint8_t pinIdx = (recvHdr->pinAddr & P_HD_TGTPIN_MASK) >> P_HD_TGTPIN_SHIFT;
+    uint32_t edgeIdx = (recvHdr->pinAddr & P_HD_DESTEDGEINDEX_MASK) 
+                        >> P_HD_DESTEDGEINDEX_SHIFT;
+    
+    // Decode the destination address
+    if(devAdr == P_ADDR_BROADCAST)
+    {   // Message is a broadcast to all devices.
+        recvDevBegin = ThreadContext->devInsts;
+        recvDevEnd = ThreadContext->devInsts + ThreadContext->numDevInsts;
+    }
+    else if (devAdr < ThreadContext->numDevInsts)
+    {   // Message is for a single device in range.
+        recvDevBegin = &ThreadContext->devInsts[devAdr];
+        recvDevEnd = recvDevBegin + 1;
+    }
+    else 
     {
-        devInst_t* device = &thr_ctxt->devInsts[cur_device];
-        // each device's OnIdle handler and if something interesting happens update the flag and run the RTS handler.
-        if (device->devType->OnIdle_Handler(thr_ctxt->properties, device))
+        handler_log(5, "dIDX OOR %d", devAdr);
+        return;    // Message target is out of range. //TODO: - log/flag
+    }
+    
+    
+    if((recvHdr->swAddr & P_SW_CNC_MASK) && (opcode > P_CNC_MAX_USER))
+    {   // Check for reserved Opcodes
+        switch(opcode)
         {
-            if (++cur_device >= thr_ctxt->numDevInsts) thr_ctxt->nextOnIdle = P_ONIDLE_CHANGE;
-            else thr_ctxt->nextOnIdle = P_ONIDLE_CHANGE | cur_device;
-            softswitch_onRTS(thr_ctxt, device);
-            return true;
+            case P_CNC_STOP:    // Stop message ends the simulation
+                ThreadContext->ctlEnd = 1;
+                break;
+#ifdef SOFTSWITCH_INSTRUMENTATION
+            case P_CNC_INSTR:   // Request Instrumentation from the Softswitch
+                ThreadContext->pendCycles = 1;
+                break;
+#endif
+            case P_CNC_INIT:
+                // **BODGE BODGE BODGE** not so temporary handler for dealing 
```

Comment:
I was worried for a second that this comment might disappear!

Reply:
It can't disappear until we get v4 as the init stuff is still crazy in v2 ;)